api/src/middleware/error-handler.ts

```diff
@@ -1,105 +1,123 @@
 				status = error.status;
 			} else if (status !== error.status) {
-				status = 500;
+				// Fallback if status has already been set by a preceding error
+				// and doesn't match the current one
+				status = FALLBACK_ERROR.status;
 			}
 
-			payload.errors.push({
+			errors.push({
 				message: error.message,
 				extensions: {
-					code: error.code,
 					...(error.extensions ?? {}),
+					// Expose error code under error's extensions data
+					code: error.code,
 				},
 			});
 
 			if (isDirectusError(error, ErrorCode.MethodNotAllowed)) {
-				res.header('Allow', (error as InstanceType<typeof MethodNotAllowedError>).extensions.allowed.join(', '));
+				res.header('Allow', error.extensions.allowed.join(', '));
 			}
 		} else {
 			logger.error(error);
 
-			status = 500;
+			status = FALLBACK_ERROR.status;
 
 			if (req.accountability?.admin === true) {
 				const localError = isObject(error) ? error : {};
-				const message = localError['message'] ?? typeof error === 'string' ? error : null;
-
-				payload = {
-					errors: [
-						{
-							message: message || 'An unexpected error occurred.',
-							extensions: {
-								code: 'INTERNAL_SERVER_ERROR',
-								...(localError['extensions'] ?? {}),
-							},
+
+				// Use 'message' prop if available, otherwise if 'error' is a string use that
+				const message =
+					(typeof localError['message'] === 'string' ? localError['message'] : null) ??
+					(typeof error === 'string' ? error : null);
+
+				errors = [
+					{
+						message: message || FALLBACK_ERROR.message,
+						extensions: {
+							code: FALLBACK_ERROR.code,
+							...(localError['extensions'] ?? {}),
 						},
-					],
-				};
+					},
+				];
 			} else {
-				payload = {
-					errors: [
-						{
-							message: 'An unexpected error occurred.',
-							extensions: {
-								code: 'INTERNAL_SERVER_ERROR',
-							},
-						},
-					],
-				};
+				// Don't expose unknown errors to non-admin users
+				errors = [{ message: FALLBACK_ERROR.message, extensions: { code: FALLBACK_ERROR.code } }];
 			}
 		}
 	}
 
-	res.status(status ?? 500);
-
-	emitter
-		.emitFilter(
-			'request.error',
-			payload.errors,
-			{},
-			{
-				database: getDatabase(),
-				schema: req.schema,
-				accountability: req.accountability ?? null,
-			},
-		)
-		.then((updatedErrors) => {
-			return res.json({ ...payload, errors: updatedErrors });
-		});
-};
+	res.status(status ?? FALLBACK_ERROR.status);
+
+	const updatedErrors = await emitter.emitFilter(
+		'request.error',
+		errors,
+		{},
+		{
+			database: getDatabase(),
+			schema: req.schema,
+			accountability: req.accountability ?? null,
+		},
+	);
 
-export default errorHandler;
+	return res.json({ errors: updatedErrors });
+});
+
+function asyncErrorHandler(
+	fn: (...args: Parameters<ErrorRequestHandler>) => Promise<any>,
+): (...args: Parameters<ErrorRequestHandler>) => Promise<ReturnType<ErrorRequestHandler>> {
+	return (err, req, res, next) =>
```

Comment:
I dont necessarily think this abstraction is an improvement over my last review but it also doesnt hurt anyone 🤷 Could we however still add back the original comment that this function requires to have 4 parameters to be marked as global error handler 😄

Reply:
What do you mean by abstraction in particular? I just removed the `Promise.resolve` wrapping in d07bed1 since the last review. Any suggestions for improvement?
IMO, the comment is no longer necessary, as it is covered by the new tests and we now (have to) forward all parameters anyway.